rfcs/text/0015_ols_phase_1.md

```diff
@@ -0,0 +1,268 @@
+- Start Date: 2020-03-01
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+---
+- [1. Summary](#1-summary)
+- [2. Motivation](#2-motivation)
+- [3. Detailed design](#3-detailed-design)
+- [4. Drawbacks](#4-drawbacks)
+- [5. Alternatives](#5-alternatives)
+- [6. Adoption strategy](#6-adoption-strategy)
+- [7. How we teach this](#7-how-we-teach-this)
+- [8. Unresolved questions](#8-unresolved-questions)
+
+# 1. Summary
+
+Object-level security ("OLS") authorizes Saved Object CRUD operations on a per-object basis.
+This RFC focuses on the [phase 1](https://github.com/elastic/kibana/issues/82725), which introduces "private" saved object types. These private types
+are owned by individual users, and are _generally_ only accessible by their owners.
+
+This RFC does not address any [followup phases](https://github.com/elastic/kibana/issues/39259), which may support sharing, and ownership of "public" objects.
+
+# 2. Motivation
+
+OLS allows saved objects to be owned by individual users. This allows Kibana to store information that is specific
+to each user, which enables further customization and collaboration throughout our solutions.
+
+The most immediate feature this unlocks is [User settings and preferences (#17888)](https://github.com/elastic/kibana/issues/17888),
+which is a very popular and long-standing request.
+
+# 3. Detailed design
+
+Phase 1 of OLS allows consumers to register "private" saved object types.
+These saved objects are owned by individual end users, and are subject to additional security controls.
+
+Public (non-private) saved object types are not impacted by this RFC. This proposal does not allow types to transition to/from `public`/`private`, and is considered out of scope for phase 1.
+
+## 3.1 Saved Objects Service
+
+### 3.1.1 Type registry
+The [saved objects type registry](https://github.com/elastic/kibana/blob/701697cc4a34d07c0508c3bdf01dca6f9d40a636/src/core/server/saved_objects/saved_objects_type_registry.ts) will allow consumers to register "private" saved object types via a new `accessClassification` property:
+
+```ts
+/**
+ * The accessClassification dictates the protection level of the saved object:
+ *  * public (default): instances of this saved object type will be accessible to all users within the given namespace, who are authorized to act on objects of this type.
+ *  * private: instances of this saved object type will belong to the user who created them, and will not be accessible by other users, except for administrators.
+ */
+export type SavedObjectsAccessClassification = 'public' | 'private';
```

Comment:
When a saved object is public it means we can quickly skip the ACL checking performance cost. But how does this fit into the longer term vision of OLS? I understand that transitioning from private to public is out of scope, but if this is something that's applied to the saved object type instead of a single object it wouldn't be possible to later share this saved object.

It feels like restricting this to a type moves us away from true OLS where the permissions are solely determined by the object itself. Is private types just technical debt until we can solve the performance problems or do we expect to always have this distinction?

Reply:
Good question. When I discussed this first phase with @arisonl, he cautioned that the future of OLS is uncertain at this point. While the [meta issue](https://github.com/elastic/kibana/issues/39259) describes a potential end state, it's not clear if/when sharing "public" objects (such as dashboards) will be possible.

Having said that, I could envision a third type of classification for objects that are sharable to other users. At that point, we could design the transition for moving public objects to become sharable. My current thinking is that `private` objects will never be sharable. They contain information that is specific to the current user, so it doesn't make sense to allow them to be shared.